### PR TITLE
Daml2ts variant support

### PR DESCRIPTION
**Pull request by shayne-fletcher-da**
Add daml2ts support for variant types (see issue https://github.com/digital-asset/daml/issues/3518).

### CODE REVIEW

File: language-support/ts/codegen/src/TsCodeGenMain.hs [hunk 1]
```diff
@@ -216,6 +222,12 @@ genDefDataType curModName tpls def = case unTypeConName (dataTypeCon def) of
             ] ++
             map ("  " <>) (onHead ("export const decoder = " <>) serDesc) ++
             ["}"]
+        genBranch ((VariantConName cons, t), i) =
+          let (typ, ser) = (genType curModName t) in
+          ( (if i == 0 then "     " else "  |  ") <> "{ tag: '" <> cons <> "'; value: " <> typ <> " }"
```
Comment (hurryabit) [inline language-support/ts/codegen/src/TsCodeGenMain.hs]: We can put the `|` even before the first case, like in OCaml. Let's do that and avoid the special case.

File: language-support/ts/codegen/src/TsCodeGenMain.hs [hunk 2]
```diff
@@ -90,6 +90,7 @@ genModule mod
         (defSers, refs) = unzip (map (genDefDataType curModName tpls) serDefs)
         header =
             ["// Generated from " <> T.intercalate "/" (unModuleName curModName) <> ".daml"
+            ,"/* eslint-disable @typescript-eslint/no-unused-vars */"
```
Comment (hurryabit) [inline language-support/ts/codegen/src/TsCodeGenMain.hs]: Do we really need this? If so, I'd be more happy to use the form which only turns off the warning for the next line whenever we need it, as we do it with the `no-namespace` warning.

Comment (shayne-fletcher-da) [reply language-support/ts/codegen/src/TsCodeGenMain.hs]: Oops, that was left over when I was testing stuff. Will remove!

File: language-support/ts/codegen/src/TsCodeGenMain.hs (hunk 1 above)
Comment (shayne-fletcher-da) [reply language-support/ts/codegen/src/TsCodeGenMain.hs]: Ooh! Testing. If all good will fix.

**Pull request merged**